benchflow-dsl/src/main/scala/cloud/benchflow/dsl/explorationspace/ExplorationSpaceGenerator.scala

```diff
@@ -63,112 +75,114 @@ object ExplorationSpaceGenerator {
 
-        ExplorationSpaceState(usersState, memoryState, environmentState)
-
-      }
-    }
-
-  }
+        // the initial block size
+        var blockSize = explorationSpaceSize
 
-  def oneAtATimeExplorationSpace(explorationSpace: ExplorationSpace): ExplorationSpaceState = {
-
-    val explorationSpaceSizeOption = calculateExplorationSpaceSize(explorationSpace)
-
-    explorationSpaceSizeOption match {
-
-      case None => ExplorationSpaceState(None, None, None)
-
-      case Some(explorationSpaceSize) => {
-
-        var shiftValue = explorationSpaceSize
-
-        val usersStateOption = explorationSpace.users.map {
+        // generate all the possible user values
+        val usersDimensionOption = explorationSpaceDimensions.users.map {
           case (list) =>
-            shiftValue = shiftValue / list.length // update shift value
-            (fillList(shiftValue, list.length, explorationSpaceSize), list.length)
+            blockSize = blockSize / list.length // update blockSize value
+            fillListWithIndices(blockSize, list.length, explorationSpaceSize)
+              .map(index => list(index))
         }
 
-        val memoryStateOption = explorationSpace.memory.map(memory => memory.map {
+        // generate all the possible memory values per service
+        val memoryDimensionOption = explorationSpaceDimensions.memory.map(memory => memory.map {
           case (serviceName, list) =>
-            shiftValue = shiftValue / list.length // update shift value
-            serviceName -> (fillList(shiftValue, list.length, explorationSpaceSize), list.length)
+            blockSize = blockSize / list.length // update blockSize value
+            serviceName -> fillListWithIndices(blockSize, list.length, explorationSpaceSize).map(index => list(index))
         })
 
-        val environmentStateOption = explorationSpace.environment.map(
+        // generate all the possible environment values per environment names and service
+        val environmentDimensionOption = explorationSpaceDimensions.environment.map(
           serviceMap => serviceMap.map {
             case (serviceName, environmentMap) => serviceName -> environmentMap.map {
               case (variableName, list) =>
-                shiftValue = shiftValue / list.length // update shift value
-                variableName -> (fillList(shiftValue, list.length, explorationSpaceSize), list.length)
+                blockSize = blockSize / list.length // update blockSize value
+                variableName -> fillListWithIndices(blockSize, list.length, explorationSpaceSize).map(index => list(index))
             }
 
           })
 
-        ExplorationSpaceState(usersState = usersStateOption, memoryStateOption, environmentStateOption)
+        ExplorationSpace(usersDimension = usersDimensionOption, memoryDimensionOption, environmentDimensionOption)
 
       }
 
     }
 
   }
 
-  def fillList(shift: Int, numValues: Int, explorationSpaceSize: Int): List[Index] = {
-
+  /**
```

Comment:
Isn't it clear from the description?
It is used to create the combinations of a given dimension (each dimension is stored in a list) to produce the cartesian product.